grandpa/primitives/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+pub const ISMP_ID: sp_runtime::ConsensusEngineId = *b"ISMP";
+
+/// GRANPA errors
+pub mod error;
+/// GRANDPA justification utilities
+pub mod justification;
+
+/// Represents a Hash in this library
+pub type Hash = H256;
+/// A commit message for this chain's block type.
+pub type Commit<H> = finality_grandpa::Commit<
+    <H as Header>::Hash,
+    <H as Header>::Number,
+    AuthoritySignature,
+    AuthorityId,
+>;
+
+/// Finality for block B is proved by providing:
+/// 1) the justification for the descendant block F;
+/// 2) headers sub-chain (B; F] if B != F;
+#[derive(Debug, PartialEq, Encode, Decode, Clone)]
+pub struct FinalityProof<H: codec::Codec> {
+    /// The hash of block F for which justification is provided.
+    pub block: Hash,
+    /// Justification of the block F.
+    pub justification: Vec<u8>,
+    /// The set of headers in the range (B; F] that we believe are unknown to the caller. Ordered.
+    pub unknown_headers: Vec<H>,
+}
+
+/// Previous light client state.
+#[derive(Debug, PartialEq, Encode, Decode, Clone)]
+pub struct ConsensusState {
+    /// Current authority set
+    pub current_authorities: AuthorityList,
+    /// Id of the current authority set.
+    pub current_set_id: u64,
+    /// latest finalized height on relay chain or standalone chain
+    pub latest_height: u32,
+    /// State machine id StateMachine::Polkadot(0) or StateMachine::Kusama(0) or
+    ///StateMachine::Grandpa(ConsensusStateId)
+    pub state_machine: StateMachine,
+    /// latest finalized height on the parachains, this map will be empty for Standalone chains
+    /// Map of para_ids
+    pub para_ids: BTreeMap<u32, bool>,
+    /// latest finalized hash on relay chain or standalone chain.
+    pub latest_hash: Hash,
+    /// slot duration for the chain
+    pub slot_duration: u64,
+}
+
+/// Holds relavant parachain proofs for both header and timestamp extrinsic.
+#[derive(Clone, Debug, Encode, Decode)]
+pub struct ParachainHeaderProofs {
+    /// State proofs that prove a parachain headers exists at a given relay chain height
+    pub state_proof: Vec<Vec<u8>>,
+    /// The parachain ids
+    pub para_ids: Vec<u32>,
+}
+
+/// Parachain headers with a Grandpa finality proof.
+#[derive(Clone, Encode, Decode)]
+pub struct ParachainHeadersWithFinalityProof<H: codec::Codec> {
+    /// The grandpa finality proof: contains relay chain headers from the
+    /// last known finalized grandpa block.
+    pub finality_proof: FinalityProof<H>,
+    /// Contains a map of relay chain header hashes to parachain headers
+    /// finalzed at the relay chain height. We check for this parachain header finalization
+    /// via state proofs. Also contains extrinsic proof for timestamp.
+    pub parachain_headers: BTreeMap<Hash, ParachainHeaderProofs>,
+}
+
+/// Hashing algorithm for the state proof
+#[derive(Debug, Encode, Decode, Clone)]
+#[cfg_attr(feature = "std", derive(serde::Deserialize, serde::Serialize))]
+pub enum HashAlgorithm {
```

Comment:
Remove these from here, move all stuff for state machine verification for both parachain and standalone chain to pallet-ismp primitives.